clients/src/main/java/org/apache/kafka/common/record/MemoryLogBufferBuilder.java

```diff
@@ -0,0 +1,476 @@
+    private final byte magic;
+    private final int initPos;
+    private final long baseOffset;
+    private final long logAppendTime;
+    private final int writeLimit;
+
+    private MemoryLogBuffer builtLogBuffer;
+    private long writtenUncompressed;
+    private long numRecords;
+    private float compressionRate;
+    private long maxTimestamp;
+    private long offsetOfMaxTimestamp;
+    private long lastOffset = -1;
+
+    public MemoryLogBufferBuilder(ByteBuffer buffer,
+                                  byte magic,
+                                  CompressionType compressionType,
+                                  TimestampType timestampType,
+                                  long baseOffset,
+                                  long logAppendTime,
+                                  int writeLimit) {
+        this.magic = magic;
+        this.timestampType = timestampType;
+        this.compressionType = compressionType;
+        this.baseOffset = baseOffset;
+        this.logAppendTime = logAppendTime;
+        this.initPos = buffer.position();
+        this.numRecords = 0;
+        this.writtenUncompressed = 0;
+        this.compressionRate = 1;
+        this.maxTimestamp = Record.NO_TIMESTAMP;
+        this.writeLimit = writeLimit;
+        this.initialBuffer = buffer;
+
+        if (compressionType != CompressionType.NONE) {
+            // for compressed records, leave space for the header and the shallow message metadata
+            // and move the starting position to the value payload offset
+            buffer.position(initPos + LogBuffer.LOG_OVERHEAD + Record.recordOverhead(magic));
+        }
+
+        // create the stream
+        bufferStream = new ByteBufferOutputStream(buffer);
+        appendStream = wrapForOutput(bufferStream, compressionType, magic, COMPRESSION_DEFAULT_BUFFER_SIZE);
+    }
+
+    public ByteBuffer initialBuffer() {
+        return initialBuffer;
+    }
+
+    public ByteBuffer buffer() {
+        return bufferStream.buffer();
+    }
+
+    public double compressionRate() {
+        return compressionRate;
+    }
+
+    /**
+     * Close this builder and return the resulting buffer.
+     * @return The built log buffer
+     */
+    public MemoryLogBuffer build() {
+        close();
+        return builtLogBuffer;
+    }
+
+    /**
+     * Get the max timestamp and its offset. If the log append time is used, then the offset will
+     * be either the first offset in the set if no compression is used or the last offset otherwise.
+     * @return The max timestamp and its offset
+     */
+    public MemoryLogBuffer.RecordsInfo info() {
+        if (timestampType == TimestampType.LOG_APPEND_TIME) {
+            long offsetOfMaxTimestamp = compressionType == CompressionType.NONE ? baseOffset : lastOffset;
+            return new MemoryLogBuffer.RecordsInfo(logAppendTime,  offsetOfMaxTimestamp);
+        }
+        return new MemoryLogBuffer.RecordsInfo(maxTimestamp, offsetOfMaxTimestamp);
```

Comment:
For compressed messageset, perhaps it's more consistent if we always return the lastOffset as offsetOfMaxTimestamp regardless of the timestamp type? We only need that for timestamp indexing. Indexing at the shallow offset level is good enough and will make the indexing logic consistent between the leader replica and the follower replica (which doesn't do decompression during log append).